File: Framework.Tests/CustomerServices/CustomerMapper.cs
using System;
using System.Collections.Generic;
using Framework.Data_Manipulation;

namespace Framework.Tests.CustomerServices
{
    class CustomerMapper : BaseMapper<Customer>
    {
        IDictionary<string, Customer> _internalData = new Dictionary<string, Customer>();
        IList<string> _sequenceDescription = new List<string>();

        public IDictionary<string, Customer> InternalData
        {
            get { return _internalData; }
        }

        public IList<string> SequenceDescription
        {
            get { return _sequenceDescription; }
        }

        public override bool Update(ref Customer entity, SuccessfulInvocationDelegate successfulInvocation,
            FailedInvocationDelegate failedInvocation)
        {
            _internalData[entity.Number] = entity;
            _sequenceDescription.Add(string.Format("{0}={1}={2}", entity.Number, "Updated", entity.GetType().FullName));
            successfulInvocation(entity);

            return true;
        }

        public override bool Delete(ref Customer entity, SuccessfulInvocationDelegate successfulInvocation,
            FailedInvocationDelegate failedInvocation)
        {
            _internalData.Remove(entity.Number);
            _sequenceDescription.Add(string.Format("{0}={1}={2}", entity.Number, "Deleted", entity.GetType().FullName));
            successfulInvocation(entity);

            return true;
        }

        public override bool Insert(ref Customer entity, SuccessfulInvocationDelegate successfulInvocation,
            FailedInvocationDelegate failedInvocation)
        {
            _internalData.Add(entity.Number, entity);
            _sequenceDescription.Add(string.Format("{0}={1}={2}", entity.Number, "Inserted", entity.GetType().Name));
            successfulInvocation(entity);

            return true;
        }
    }
}

File: DotNet/C_PlusPlus/Framework.Tests/CustomerMapper.cpp
#include "stdafx.h"

#include <algorithm>

#include "Customer.h"
#include "CustomerMapper.h"
#include "ContainerUtility.h"

using namespace CustomerServices;

class CustomerMapper::Implementation
{
private:
	/*
	Reference Wrapper: Elements referred to in a container
	 exists as long as the container exists
	*/
	typedef unordered_map<string, Customer*> Dictionary;
private:
	Dictionary m_InternalData;

	void CleanMapResources(Dictionary& dictionary)
	{
		for_each(dictionary.begin(), dictionary.end(),
			[](pair<const string, Customer*> &element) {

			/*Delete object*/
			delete element.second;
			element.second = NULL;
			/******************/
		});

		MapRemoveValues(dictionary, [](pair<const string, Customer*> element) {
			return (element.second == NULL);
		});
	}
public:
	Implementation() :
		m_InternalData()
	{ }
	Implementation(Implementation&&) = default;
	Implementation& operator=(Implementation&&) = default;
	Implementation(const Implementation&) = default;
	Implementation& operator=(const Implementation&) = default;

	~Implementation()
	{
		//Only objects instantiated within this class are to be destroyed
		/*Alternatively, variation of Item-33*/
		CleanMapResources(m_InternalData);
		/*************************************/
		/*************************************/
	}

	void AddEditCustomer(Customer& customer)
	{
		EfficientAddOrUpdateByPtr(m_InternalData, customer.GetNumber(), &customer);
	}

	size_t CollectionCount() const
	{
		return m_InternalData.size();
	}

	Customer* GetCustomer(const string& customerNumber) const
	{
		Customer **retVal = GetValue(m_InternalData, customerNumber);

		return (retVal == NULL) ? NULL : (Customer*)(*retVal);
	}
};

CustomerMapper::CustomerMapper() :
	BaseMapperConcrete<Customer>(),
	pImpl{ make_unique<Implementation>() } { }

CustomerMapper::~CustomerMapper() {}

const string CustomerMapper::CUST_NO("Customer Number");
const string CustomerMapper::CUST_NAME("Customer Name");
const string CustomerMapper::OPERATION("Operation");
const string CustomerMapper::COLLECTION_COUNT("Collection Count");

bool CustomerMapper::ConcreteInsert(Customer* entity, const SuccessfulInvocationDelegate* successfulInvocation, const FailedInvocationDelegate* failedInvocation)
{
	BaseMapperHashtable results;

	if (entity == NULL)
	{
		if(failedInvocation != NULL) (*failedInvocation)(entity, &results);

		return false;
	}

	if (pImpl) pImpl->AddEditCustomer(*entity);

	EfficientAddOrUpdate(results, CUST_NO, any(entity->GetNumber()));
	EfficientAddOrUpdate(results, CUST_NAME, any(entity->GetName()));
	EfficientAddOrUpdate(results, OPERATION, any(string("Insert")));
	EfficientAddOrUpdate(results, COLLECTION_COUNT, any(pImpl->CollectionCount()));

	if(successfulInvocation != NULL) (*successfulInvocation)(entity, &results);

	return true;
}

bool CustomerMapper::ConcreteUpdate(Customer* entity, const SuccessfulInvocationDelegate* successfulInvocation, const FailedInvocationDelegate* failedInvocation)
{
	BaseMapperHashtable results;

	if (entity == NULL)
	{
		if (failedInvocation != NULL) (*failedInvocation)(entity, &results);

		return false;
	}

	EfficientAddOrUpdate(results, OPERATION, any(string("Update")));

	if (successfulInvocation != NULL) (*successfulInvocation)(entity, &results);

	return true;
}

bool CustomerMapper::ConcreteDelete(Customer* entity, const SuccessfulInvocationDelegate* successfulInvocation, const FailedInvocationDelegate* failedInvocation)
{

	BaseMapperHashtable results;

	if (entity == NULL)
	{
		if (failedInvocation != NULL) (*failedInvocation)(entity, &results);

		return false;
	}

	EfficientAddOrUpdate(results, OPERATION, any(string("Delete")));

	if (successfulInvocation != NULL) (*successfulInvocation)(entity, &results);

	return true;
}

size_t CustomerMapper::GetCollectionCount() const
{
	return (!pImpl) ? 0 : pImpl->CollectionCount();
}

Customer* CustomerMapper::GetCustomer(const string& customerNumber) const
{
	return (!pImpl) ? NULL : pImpl->GetCustomer(customerNumber);
}